### PR TITLE
[Storage] Add AptosDBError

### PR DESCRIPTION
**Pull request by areshand**
### Description

Introduce AptosDbError to replace anyhow Error.
The goal is to have custom error codes (eg: differentiate not found error from other internal errors.)

### Test Plan
Existing test.

### CODE REVIEW

File: storage/storage-interface/src/lib.rs [hunk 1]
```diff
@@ -53,9 +54,10 @@ use crate::state_delta::StateDelta;
 use aptos_scratchpad::SparseMerkleTree;
 pub use executed_trees::ExecutedTrees;
 
+pub type Result<T, E = AptosDbError> = std::result::Result<T, E>;
 // This is last line of defense against large queries slipping through external facing interfaces,
 // like the API and State Sync, etc.
-pub const MAX_REQUEST_LIMIT: u64 = 20_000;
+pub const MAX_REQUEST_LIMIT: u64 = 10000;
```
[the remark this answers is not in the document]
Comment (areshand) [reply storage/storage-interface/src/lib.rs:60]: accidental change

File: storage/aptosdb/src/event_store/mod.rs [hunk 2]
```diff
@@ -511,19 +510,19 @@ impl<'a> EventHashReader<'a> {
 }
 
 impl<'a> HashReader for EventHashReader<'a> {
-    fn get(&self, position: Position) -> Result<HashValue> {
+    fn get(&self, position: Position) -> Result<HashValue, anyhow::Error> {
         self.store
             .event_db
             .get::<EventAccumulatorSchema>(&(self.version, position))?
-            .ok_or_else(|| format_err!("Hash at position {:?} not found.", position))
+            .ok_or_else(|| anyhow!(format!("Hash at position {:?} not found.", position)))
```
Comment (msmouse) [inline storage/aptosdb/src/event_store/mod.rs:517]: `format!` not necessary

File: state-sync/storage-service/server/src/tests/mock.rs [hunk 3]
```diff
@@ -231,40 +231,40 @@ mock! {
             &self,
             start_epoch: u64,
             end_epoch: u64,
-        ) -> Result<EpochChangeProof>;
+        ) -> aptos_storage_interface::Result<EpochChangeProof>;
```
Comment (msmouse) [inline state-sync/storage-service/server/src/tests/mock.rs:234]: import it

File: state-sync/storage-service/server/src/storage.rs [hunk 4]
```diff
@@ -2,10 +2,9 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::{error::Error, metrics::increment_network_frame_overflow};
-use anyhow;
 use aptos_config::config::StorageServiceConfig;
 use aptos_logger::debug;
-use aptos_storage_interface::DbReader;
+use aptos_storage_interface::{AptosDbError, DbReader, Result as StorageResult};
```
Comment (msmouse) [inline state-sync/storage-service/server/src/storage.rs:7]: okay.. this works as well..

File: state-sync/state-sync-driver/src/tests/mocks.rs [hunk 5]
```diff
@@ -140,34 +140,34 @@ mock! {
             txn_list_with_proof: TransactionListWithProof,
             verified_target_li: &LedgerInfoWithSignatures,
             epoch_change_li: Option<&'a LedgerInfoWithSignatures>,
-        ) -> Result<()>;
+        ) -> AnyhowResult<()>;
```
Comment (msmouse) [inline state-sync/state-sync-driver/src/tests/mocks.rs:143]: `anyhow::Result` is not too much longer, and I think explicitly say DbResult is better than this.

File: peer-monitoring-service/server/src/tests.rs [hunk 6]
```diff
@@ -646,130 +647,135 @@ pub fn create_mock_db_reader() -> MockDatabaseReader {
...
-            version: Version,
-            start_idx: usize,
-            chunk_size: usize,
-        ) -> anyhow::Result<StateValueChunkWithProof>;
-
-        fn get_epoch_snapshot_prune_window(&self) -> anyhow::Result<usize>;
+mod database_mock {
+    use super::*;
+    type Result<T, E = AptosDbError> = std::result::Result<T, E>;
+
```
Comment (msmouse) [inline peer-monitoring-service/server/src/tests.rs:653]: define DbResult and importing it looks better

File: consensus/src/quorum_store/quorum_store_builder.rs [hunk 7]
```diff
@@ -392,6 +392,7 @@ impl InnerBuilder {
                     match aptos_db_clone.get_latest_ledger_info() {
                         Ok(ledger_info) => BatchResponse::NotFound(ledger_info),
                         Err(e) => {
+                            let e = anyhow::Error::from(e);
```
Comment (msmouse) [inline consensus/src/quorum_store/quorum_store_builder.rs:395]: why not derive `Debug` for our errors?

**Review by msmouse — APPROVED** [review]

File: storage/storage-interface/src/lib.rs [hunk 8]
```diff
@@ -463,8 +466,9 @@ pub trait DbReader: Send + Sync {
 
     /// Returns the latest ledger info.
     fn get_latest_ledger_info(&self) -> Result<LedgerInfoWithSignatures> {
-        self.get_latest_ledger_info_option()
-            .and_then(|opt| opt.ok_or_else(|| format_err!("Latest LedgerInfo not found.")))
+        self.get_latest_ledger_info_option().and_then(|opt| {
+            opt.ok_or_else(|| AptosDbError::Other("Latest LedgerInfo not found.".to_string()))
```
[the remark this answers is not in the document]
Comment (areshand) [reply storage/storage-interface/src/lib.rs:470]: good to know. next time

File: storage/storage-interface/src/errors.rs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+impl From<AptosDbError> for StateviewError {
+    fn from(error: AptosDbError) -> Self {
+        match error {
+            AptosDbError::NotFound(msg) => StateviewError::NotFound(msg),
+            AptosDbError::Other(msg) => StateviewError::Other(msg),
+            _ => StateviewError::Other(format!("{}", error)),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (areshand) [reply storage/storage-interface/src/errors.rs:80]: yes, allowed to define here

File: storage/schemadb/tests/db.rs [hunk 10]
```diff
@@ -178,7 +179,7 @@ fn collect_values<S: Schema>(db: &TestDB) -> Vec<(S::Key, S::Value)> {
         .iter::<S>(Default::default())
         .expect("Failed to create iterator.");
     iter.seek_to_first();
-    iter.collect::<Result<Vec<_>>>().unwrap()
+    iter.collect::<Result<Vec<_>, AptosDbError>>().unwrap()
```
[the remark this answers is not in the document]
Comment (areshand) [reply storage/schemadb/tests/db.rs:182]: compiler asks for explicitly define

File: consensus/src/quorum_store/quorum_store_builder.rs (hunk 7 above)
Comment (areshand) [reply consensus/src/quorum_store/quorum_store_builder.rs:395]: Not sure how debug helps here.

**Review by grao1991 — APPROVED** [review]

File: consensus/src/quorum_store/quorum_store_builder.rs (hunk 7 above)
Comment (msmouse) [reply consensus/src/quorum_store/quorum_store_builder.rs:395]: then `error = ?e` can render? otherwise why are you defining `e` here?

**Pull request merged**